Add annapurna workflows and dependencies (from fork) (#17)
eProsima/eProsima-CI: README.md

# eProsima-CI

This is a github CI auxiliary repo that contains common github steps, actions, and workflows shared along multiple eProsima libraries and tools.
The main idea is to collect every repeated or generic step of any CI to have a single more maintainable generic point.

---

## Index

- [eProsima-CI](#eprosima-ci)

>   - [Index](#index)

Reviewer: I do not think these two should be in the index. Usually the table of contents does not include the title nor the table of contents itself.

Author: The index is autogenerated with an extension of vscode, so every time I save the file it autogenerates all. The extension allows to remove ##index but not the main one. Is it too much trouble to leave it this way?

  - [User manual](#user-manual)
  - [Out-of-the-box Actions Implemented](#out-of-the-box-actions-implemented)
  - [Actions Implemented](#actions-implemented)
    - [Multiplatform](#multiplatform)
    - [Ubuntu](#ubuntu)
  - [Dependencies built](#dependencies-built)
    - [Artifacts uploaded](#artifacts-uploaded)
    - [Generate artifacts manually](#generate-artifacts-manually)

---

## User manual

> In order to use this repository, we recommend you:

Reviewer: Please, avoid the personal style. Apply this suggestion to other parts in this README.

```suggestion
In order to use this repository, the following recommendations are suggested:
```

> 1. **Always use a fix version**. When using any of these actions, set it with a `@v<X>` version.

Reviewer: Please, break line after each sentence.

Following this recommendation, I suppose that a release policy should be stated here on the README.

2. Use `multiplatform` actions when possible. They automatically will redirect the action to the proper OS.
3. If you want a long well preserved CI, use `@vX.Y` versioning, to be sure that CI will not change and break something in the future. However, this may be a problem because it will not be updated with improvements or external versions.
4. Actions are documented in their own `<platform>/<action_name>/action.yml` files. Check them for description and argument definitions.

---

## Out-of-the-box Actions Implemented

- [uncrustify](ubuntu/uncrustify/action.yml)
  - Check the C++ linter of the new files added to the repository.
  - *Only on ubuntu*.

- [python_linter](ubuntu/python_linter/action.yml)
  - Check the Python linter of the new files added to the repository.
  - *Only on ubuntu*.

- [sphinx_docs](ubuntu/sphinx_docs/action.yml)
  - Build Sphinx documentation, test it and upload results.
  - *Only on ubuntu*.

>   - *Only for documentation projects based in `cmake_utils`*.

Reviewer: :sweat_smile: sorry for being ambiguous myself. I was referring to the `cmake_utils` repository, and not to the Sphinx one.

  - Download a `.repos` and a `colcon.meta` file from a repository.

- [get_file_from_repo](multiplatform/get_file_from_repo/action.yml)
  - Download a file from a repository.

- [get_workflow_id](multiplatform/get_workflow_id/action.yml)
  - Get the ID of a specific workflow run.

- [install_colcon](multiplatform/install_colcon/action.yml)
  - Install colcon and its dependencies depending on the platform.

- [install_fastdds_dependencies](multiplatform/install_fastdds_dependencies/action.yml)
  - Install Fast DDS and general eProsima projects dependencies.

- [install_gtest](multiplatform/install_gtest/action.yml)
  - Instal `gtest` and `gmock` C++ library.

- [install_openssl](multiplatform/install_openssl/action.yml)
  - Instal `Open SSL` C++ library.

- [install_yamlcpp](multiplatform/install_yamlcpp/action.yml)
  - Instal `yaml-cpp` C++ library.

- [tsan_build_test](multiplatform/tsan_build_test/action.yml)
  - Build a project using colcon and TSAN flags and execute tests.

- [vcs_import](multiplatform/vcs_import/action.yml)
  - Import several repositories from a Yaml file using `vcs`.

### Ubuntu

- [get_git_diff_files](ubuntu/get_git_diff_files/action.yml)

>   - Get the files that differ from one github reference to another. Can grep result.

Reviewer: ```suggestion
  - Get the files that differ from one github reference to another. The result can be parsed with `grep`.
```

- [git_fetch_all](ubuntu/git_fetch_all/action.yml)
  - Get all branches of a specific repository.

- [install_apt_packages](ubuntu/install_apt_packages/action.yml)
  - Install apt packages.

- [install_python_packages](ubuntu/install_python_packages/action.yml)
  - Install python packages.

- [set_platform](ubuntu/set_platform/action.yml)
  - Set the platform OS version in a environment variable.

---

## Dependencies built

There are several workflows implemented that build projects and upload them as artifacts.
These are used for other projects to speed up the built process of the dependencies.

### Artifacts uploaded

> So far, there are 3 workflows running to upload artifacts:

Reviewer: ```suggestion
So far, the following workflows are running to upload artifacts:
```

Instead of having to change the number each time a new workflow is added, the developer only has to add the workflow to the list.

- [Fast DDS](<https://github.com/eProsima/Fast-DDS>)
- [dev-utils](<https://github.com/eProsima/dev-utils>)
- [DDS Pipe](<https://github.com/eProsima/DDS-Pipe/>)

> These are the artifacts that are generated every night with the latest versions of each project:

Reviewer: Maybe a reference or explanation about which configurations options are enabled and also how to introduce some other artifact with other configuration parameters would be nice to have.

Author: :sweat:

  - built_fastdds_ubuntu-22.04_Debug_nightly
  - built_fastdds_ubuntu-22.04_Release_nightly
  - built_fastdds_windows-2019_Debug_nightly
  - built_fastdds_windows-2019_Release_nightly
  - built_fastdds_windows-2022_Debug_nightly
  - built_fastdds_windows-2022_Release_nightly

- dev-utils
  - built_dev_utils_ubuntu-20.04_Debug_nightly
  - built_dev_utils_ubuntu-20.04_Release_nightly
  - built_dev_utils_ubuntu-22.04_Debug_nightly
  - built_dev_utils_ubuntu-22.04_Release_nightly
  - built_dev_utils_windows-2019_Debug_nightly
  - built_dev_utils_windows-2019_Release_nightly
  - built_dev_utils_windows-2022_Debug_nightly
  - built_dev_utils_windows-2022_Release_nightly

- DDS Pipe
  - built_ddspipe_ubuntu-20.04_Debug_nightly
  - built_ddspipe_ubuntu-20.04_Release_nightly
  - built_ddspipe_ubuntu-22.04_Debug_nightly
  - built_ddspipe_ubuntu-22.04_Release_nightly
  - built_ddspipe_windows-2019_Debug_nightly
  - built_ddspipe_windows-2019_Release_nightly
  - built_ddspipe_windows-2022_Debug_nightly
  - built_ddspipe_windows-2022_Release_nightly

In order to use one of these artifacts, use the following action as a step:

```yml
- name: Get fastdds artifact

>   uses: jparisu/eProsima-CI/multiplatform/download_dependency@main

Reviewer: Maybe insted of referring to `main`, this should say `vX.Y`. Even if it is not possible to directly copy the snippet, the user would have to set the proper version. Otherwise it could be overlooked.

Author: I left it as this so the last commit that updates all links changes also this one

Reviewer: Ok. But I was not referring to the repository itself but to the version of the action. I will leave it pending.

Author: If I understand correctly, you suggest to write `jparisu/eProsima-CI/multiplatform/download_dependency@vX.Y`.
I suggest to leave it this way so every time a new version is going to be updated, you can find everywhere for `@main` and set it to `@vX.Y`, and this includes this, so it will change along the rest.

  with:
    artifact_name: built_fastdds_ubuntu-20.04_Debug_nightly
    workflow_source: build_fastdds.yml
    workflow_source_repository: eProsima/eProsima-CI
    # Path where to download the installs from the artifact
    target_workspace: ${{ github.workspace }}/install
    # If inside an action, this value must be generated in workflow and passed as argument
    secret_token: ${{ secrets.GITHUB_TOKEN }}
```

### Generate artifacts manually

Artifacts could be executed manually with different dependencies and/or different CMake arguments.

> In order to do so there are 2 possibilities:

Reviewer: I see that there is still some documentation pending

# Fast-DDS-CI

> This is the eProsima Annapurna Team branch of eprosima-CI

This is a github CI auxiliary repo that contains common github steps and actions shared along multiple eProsima libraries and tools.
The main idea is to collect every repeated or generic step of any CI to have a single more maintainable generic point.

> ## Steps Implemented

Author: I'll leave this to a future commit with documentation changes. Do not resolve conversation

* **Ubuntu (generic)**

  * `set_platform`
    * Set the platform OS version in a environment variable.

  * `install_apt_packages`
    * Install generic apt packages required by these actions and most eProsima projects.

  * `install_python_packages`
    * Install generic python packages required by these actions and most eProsima projects.

  * `install_colcon`
    * Install colcon and its dependencies depending on the platform.

  * `get_git_diff_files`
    * Get the files that differ from one github reference to another. Can grep result.

  * `git_fetch_all`
    * Get all branches of a specific repository.

> ## Full Actions Implemented

Author: Finally I did it in a way that could be run from ubuntu 20 and 22.
I think it is better than having it in `ubuntu` folder and not be able to run it from one of them.

> ## Full Actions Implemented

Reviewer: Maybe, in order to not pollute the root directory, I suggest moving the `install_uncrustify` actions inside the `ubuntu` folder. Something like the structure shown in the following tree:

```
ubuntu
|
|- ubuntu-20.04
|   |- install_uncrustify
|
|- ubuntu-22.04
|   |- install_uncrustify
```

> ## Full Actions Implemented

Author: I am not sure if I like that; ubuntu-20 and ubuntu-22 would be a bit lost between all actions...
At least I remove `ubuntu`-20.04 from the name, so folders are 20.04 and 22.04 and are more distinguishable

> ## Full Actions Implemented

Reviewer: I agree with your criteria of naming the folders with the version directly so they can be easily found.
